From 3998f45bfea096436466b29aeef194903f9605a8 Mon Sep 17 00:00:00 2001
From: Tim Hayes <tim.hayes@smartrg.com>
Date: Wed, 7 Oct 2020 07:53:03 -0700
Subject: [PATCH] OWRT-5535: net: dsa: mt7530: add get_stats

add get_stats64 to update netdev stats using RMON counters

Signed-off-by: Tim Hayes <tim.hayes@smartrg.com>
---
 drivers/net/dsa/mt7530.c |  202 +++++++++++++++++++++++++++++--------
 drivers/net/dsa/mt7530.h |   17 ---
 2 files changed, 165 insertions(+), 54 deletions(-)

--- a/drivers/net/dsa/mt7530.c
+++ b/drivers/net/dsa/mt7530.c
@@ -30,17 +30,95 @@ static struct mt753x_pcs *pcs_to_mt753x_
 	return container_of(pcs, struct mt753x_pcs, pcs);
 }
 
+typedef void (*mt7530_mib_stats_fn)(struct rtnl_link_stats64 *, u64);
+
+struct mt7530_mib_desc {
+	unsigned int size;
+	unsigned int offset;
+	mt7530_mib_stats_fn stats;
+	const char *name;
+};
+
+#define MIB_DESC(_s, _o, _fn, _n)	\
+{			\
+	.size = (_s),	\
+	.offset = (_o),	\
+	.stats = (_fn), \
+	.name = (_n),	\
+}
+
+static inline void mt7530_update_tx_bytes(struct rtnl_link_stats64 *s, u64 v) { s->tx_bytes += v; }
+static inline void mt7530_update_tx_dropped(struct rtnl_link_stats64 *s, u64 v) { s->tx_dropped += v; }
+static inline void mt7530_update_tx_packets(struct rtnl_link_stats64 *s, u64 v) { s->tx_packets += v; }
+static inline void mt7530_update_tx_errors(struct rtnl_link_stats64 *s, u64 v) { s->tx_errors += v; }
+static inline void mt7530_update_collisions(struct rtnl_link_stats64 *s, u64 v) { s->collisions += v; }
+static inline void mt7530_update_rx_bytes(struct rtnl_link_stats64 *s, u64 v) { s->rx_bytes += v; }
+static inline void mt7530_update_rx_dropped(struct rtnl_link_stats64 *s, u64 v) { s->rx_dropped += v; }
+static inline void mt7530_update_rx_packets(struct rtnl_link_stats64 *s, u64 v) { s->rx_packets += v; }
+static inline void mt7530_update_rx_errors(struct rtnl_link_stats64 *s, u64 v) { s->rx_errors += v; }
+static inline void mt7530_update_rx_crc_errors(struct rtnl_link_stats64 *s, u64 v) { s->rx_crc_errors += v; }
+
+static inline void mt7530_update_rx_length_errors(struct rtnl_link_stats64 *s, u64 v)
+{
+	s->rx_errors += v;
+	s->rx_length_errors += v;
+}
+
+static inline void mt7530_update_tx_multicast(struct rtnl_link_stats64 *s, u64 v)
+{
+	s->multicast += v;
+	s->tx_packets += v;
+}
+
+static inline void mt7530_update_rx_multicast(struct rtnl_link_stats64 *s, u64 v)
+{
+	s->multicast += v;
+	s->rx_packets += v;
+}
+
 /* String, offset, and register size in bytes if different from 4 bytes */
 static const struct mt7530_mib_desc mt7530_mib[] = {
-	MIB_DESC(1, MT7530_PORT_MIB_TX_DROP, "TxDrop"),
-	MIB_DESC(1, MT7530_PORT_MIB_TX_CRC_ERR, "TxCrcErr"),
-	MIB_DESC(1, MT7530_PORT_MIB_TX_COLLISION, "TxCollision"),
-	MIB_DESC(1, MT7530_PORT_MIB_RX_DROP, "RxDrop"),
-	MIB_DESC(1, MT7530_PORT_MIB_RX_FILTERING, "RxFiltering"),
-	MIB_DESC(1, MT7530_PORT_MIB_RX_CRC_ERR, "RxCrcErr"),
-	MIB_DESC(1, MT7530_PORT_MIB_RX_CTRL_DROP, "RxCtrlDrop"),
-	MIB_DESC(1, MT7530_PORT_MIB_RX_INGRESS_DROP, "RxIngressDrop"),
-	MIB_DESC(1, MT7530_PORT_MIB_RX_ARL_DROP, "RxArlDrop"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_DROP, mt7530_update_tx_dropped, "TxDrop"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_CRC_ERR, mt7530_update_tx_errors, "TxCrcErr"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_UNICAST, mt7530_update_tx_packets, "TxUnicast"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_MULTICAST, mt7530_update_tx_multicast, "TxMulticast"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_BROADCAST, mt7530_update_tx_packets, "TxBroadcast"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_COLLISION, mt7530_update_collisions, "TxCollision"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_SINGLE_COLLISION, mt7530_update_collisions, "TxSingleCollision"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_MULTIPLE_COLLISION, mt7530_update_collisions, "TxMultipleCollision"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_DEFERRED, NULL, "TxDeferred"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_LATE_COLLISION, mt7530_update_collisions, "TxLateCollision"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_EXCESSIVE_COLLISION, mt7530_update_collisions, "TxExcessiveCollistion"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_PAUSE, NULL, "TxPause"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_PKT_SZ_64, NULL, "TxPktSz64"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_PKT_SZ_65_TO_127, NULL, "TxPktSz65To127"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_PKT_SZ_128_TO_255, NULL, "TxPktSz128To255"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_PKT_SZ_256_TO_511, NULL, "TxPktSz256To511"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_PKT_SZ_512_TO_1023, NULL, "TxPktSz512To1023"),
+	MIB_DESC(1, MT7530_PORT_MIB_TX_PKT_SZ_1024_TO_MAX, NULL, "Tx1024ToMax"),
+	MIB_DESC(2, MT7530_PORT_MIB_TX_BYTES, mt7530_update_tx_bytes, "TxBytes"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_DROP, mt7530_update_rx_dropped, "RxDrop"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_FILTERING, mt7530_update_rx_dropped, "RxFiltering"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_UNICAST, mt7530_update_rx_packets, "RxUnicast"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_MULTICAST, mt7530_update_rx_multicast, "RxMulticast"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_BROADCAST, mt7530_update_rx_packets, "RxBroadcast"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_ALIGN_ERR, mt7530_update_rx_errors, "RxAlignErr"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_CRC_ERR, mt7530_update_rx_crc_errors, "RxCrcErr"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_UNDER_SIZE_ERR, mt7530_update_rx_length_errors, "RxUnderSizeErr"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_FRAG_ERR, mt7530_update_rx_errors, "RxFragErr"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_OVER_SZ_ERR, mt7530_update_rx_errors, "RxOverSzErr"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_JABBER_ERR, mt7530_update_rx_errors, "RxJabberErr"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_PAUSE, NULL, "RxPause"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_PKT_SZ_64, NULL, "RxPktSz64"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_PKT_SZ_65_TO_127, NULL, "RxPktSz65To127"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_PKT_SZ_128_TO_255, NULL, "RxPktSz128To255"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_PKT_SZ_256_TO_511, NULL, "RxPktSz256To511"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_PKT_SZ_512_TO_1023, NULL, "RxPktSz512To1023"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_PKT_SZ_1024_TO_MAX, NULL, "RxPktSz1024ToMax"),
+	MIB_DESC(2, MT7530_PORT_MIB_RX_BYTES, mt7530_update_rx_bytes, "RxBytes"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_CTRL_DROP, mt7530_update_rx_dropped, "RxCtrlDrop"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_INGRESS_DROP, mt7530_update_rx_dropped, "RxIngressDrop"),
+	MIB_DESC(1, MT7530_PORT_MIB_RX_ARL_DROP, mt7530_update_rx_dropped, "RxArlDrop"),
 };
 
 static void
@@ -770,6 +848,72 @@ mt7530_get_ethtool_stats(struct dsa_swit
 	}
 }
 
+static void
+mt7530_mib_fetch_port_stat(struct mt7530_priv *priv, int port,
+			 struct rtnl_link_stats64 *storage)
+{
+	const u32 base_reg = MT7530_PORT_MIB_COUNTER(port);
+	u32 i;
+
+	for (i = 0; i < ARRAY_SIZE(mt7530_mib); i++) {
+		const struct mt7530_mib_desc *mib = &mt7530_mib[i];
+		u64 val;
+
+		/* skip counters with no link_stats mapping */
+		if (!mib->stats)
+			continue;
+
+		val = mt7530_read(priv, base_reg + mib->offset);
+		if (mib->size == 2) {
+			u64 hi = mt7530_read(priv, base_reg + mib->offset + 4);
+			val |= hi << 32;
+		}
+
+		mib->stats(storage, val);
+	}
+}
+
+static void
+mt7530_mib_work_func(struct work_struct *work)
+{
+	struct rtnl_link_stats64 stats[MT7530_NUM_PORTS] = {0};
+	struct mt7530_priv *priv;
+	int port;
+
+	priv = container_of(work, struct mt7530_priv, mib_work.work);
+
+	for (port = 0; port < MT7530_NUM_PORTS; port++)
+		mt7530_mib_fetch_port_stat(priv, port, &stats[port]);
+
+	spin_lock(&priv->stats_lock);
+	memcpy(priv->stats, stats, sizeof(stats));
+	spin_unlock(&priv->stats_lock);
+
+	schedule_delayed_work(&priv->mib_work,
+			      msecs_to_jiffies(MT7530_MIB_WORK_DELAY));
+}
+
+static void
+mt7530_get_stats64(struct dsa_switch *ds, int port,
+			struct rtnl_link_stats64 *stats)
+{
+	struct mt7530_priv *priv = ds->priv;
+
+	if (!stats) {
+		dev_err(ds->dev, "NULL stats64 pointer\n");
+		return;
+	}
+
+	spin_lock(&priv->stats_lock);
+	memcpy(stats, &priv->stats[port], sizeof(struct rtnl_link_stats64));
+	spin_unlock(&priv->stats_lock);
+
+	if (!priv->mib_work_started) {
+		priv->mib_work_started = true;
+		schedule_delayed_work(&priv->mib_work, 0);
+	}
+}
+
 static int
 mt7530_get_sset_count(struct dsa_switch *ds, int port, int sset)
 {
@@ -888,51 +1032,6 @@ static void mt7530_get_rmon_stats(struct
 	*ranges = mt7530_rmon_ranges;
 }
 
-static void mt7530_get_stats64(struct dsa_switch *ds, int port,
-			       struct rtnl_link_stats64 *storage)
-{
-	struct mt7530_priv *priv = ds->priv;
-	uint64_t data;
-
-	/* MIB counter doesn't provide a FramesTransmittedOK but instead
-	 * provide stats for Unicast, Broadcast and Multicast frames separately.
-	 * To simulate a global frame counter, read Unicast and addition Multicast
-	 * and Broadcast later
-	 */
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_RX_UNICAST, 1,
-			       &storage->rx_packets);
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_RX_MULTICAST, 1,
-			       &storage->multicast);
-	storage->rx_packets += storage->multicast;
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_RX_BROADCAST, 1,
-			       &data);
-	storage->rx_packets += data;
-
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_TX_UNICAST, 1,
-			       &storage->tx_packets);
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_TX_MULTICAST, 1,
-			       &data);
-	storage->tx_packets += data;
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_TX_BROADCAST, 1,
-			       &data);
-	storage->tx_packets += data;
-
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_RX_BYTES, 2,
-			       &storage->rx_bytes);
-
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_TX_BYTES, 2,
-			       &storage->tx_bytes);
-
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_RX_DROP, 1,
-			       &storage->rx_dropped);
-
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_TX_DROP, 1,
-			       &storage->tx_dropped);
-
-	mt7530_read_port_stats(priv, port, MT7530_PORT_MIB_RX_CRC_ERR, 1,
-			       &storage->rx_crc_errors);
-}
-
 static void mt7530_get_eth_ctrl_stats(struct dsa_switch *ds, int port,
 				      struct ethtool_eth_ctrl_stats *ctrl_stats)
 {
@@ -3261,11 +3360,11 @@ const struct dsa_switch_ops mt7530_switc
 	.preferred_default_local_cpu_port = mt753x_preferred_default_local_cpu_port,
 	.get_strings		= mt7530_get_strings,
 	.get_ethtool_stats	= mt7530_get_ethtool_stats,
+	.get_stats64		= mt7530_get_stats64,
 	.get_sset_count		= mt7530_get_sset_count,
 	.get_eth_mac_stats	= mt7530_get_eth_mac_stats,
 	.get_rmon_stats		= mt7530_get_rmon_stats,
 	.get_eth_ctrl_stats	= mt7530_get_eth_ctrl_stats,
-	.get_stats64		= mt7530_get_stats64,
 	.set_ageing_time	= mt7530_set_ageing_time,
 	.port_enable		= mt7530_port_enable,
 	.port_disable		= mt7530_port_disable,
@@ -3397,6 +3496,11 @@ mt7530_probe_common(struct mt7530_priv *
 	mutex_init(&priv->reg_mutex);
 	dev_set_drvdata(dev, priv);
 
+	spin_lock_init(&priv->stats_lock);
+	INIT_DELAYED_WORK(&priv->mib_work, mt7530_mib_work_func);
+	priv->mib_work_started = false;
+	memset(priv->stats, 0, sizeof(priv->stats));
+
 	return 0;
 }
 EXPORT_SYMBOL_GPL(mt7530_probe_common);
@@ -3404,6 +3508,8 @@ EXPORT_SYMBOL_GPL(mt7530_probe_common);
 void
 mt7530_remove_common(struct mt7530_priv *priv)
 {
+	cancel_delayed_work_sync(&priv->mib_work);
+
 	if (priv->irq_domain)
 		mt7530_free_mdio_irq(priv);
 
--- a/drivers/net/dsa/mt7530.h
+++ b/drivers/net/dsa/mt7530.h
@@ -768,18 +768,7 @@ enum mt7531_xtal_fsel {
 #define  REG_GSWCK_EN			BIT(0)
 #define  REG_TRGMIICK_EN		BIT(1)
 
-#define MIB_DESC(_s, _o, _n)	\
-	{			\
-		.size = (_s),	\
-		.offset = (_o),	\
-		.name = (_n),	\
-	}
-
-struct mt7530_mib_desc {
-	unsigned int size;
-	unsigned int offset;
-	const char *name;
-};
+#define MT7530_MIB_WORK_DELAY	1000 /* msecs */
 
 struct mt7530_fdb {
 	u16 vid;
@@ -900,6 +889,10 @@ struct mt7530_priv {
 	int (*create_sgmii)(struct mt7530_priv *priv);
 	u8 active_cpu_ports;
 	struct mdio_device *mdiodev;
+	spinlock_t stats_lock;
+	struct delayed_work mib_work;
+	bool mib_work_started;
+	struct rtnl_link_stats64 stats[MT7530_NUM_PORTS];
 };
 
 struct mt7530_hw_vlan_entry {
